pkg/utils/utils_mlx.go

```diff
@@ -0,0 +1,116 @@
+	glog.Infof("ParseMstconfigOutput(): Attributes %v", attributes)
+	fwCurrent = map[string]string{}
+	fwNext = map[string]string{}
+	formatRegex := regexp.MustCompile(`(?P<Attribute>\w+)\s+(?P<Default>\S+)\s+(?P<Current>\S+)\s+(?P<Next>\S+)`)
+	mstOutputLines := strings.Split(mstOutput, "\n")
+	for _, attr := range attributes {
+		for _, line := range mstOutputLines {
+			if strings.Contains(line, attr) {
+				regexResult := formatRegex.FindStringSubmatch(line)
+				fwCurrent[attr] = regexResult[3]
+				fwNext[attr] = regexResult[4]
+				break
+			}
+		}
+	}
+	return
+}
+
+func mellanoxBlueFieldMode(PciAddress string) (BlueField2Mode, error) {
+	glog.V(2).Infof("MellanoxBlueFieldMode():checking mode for card %s", PciAddress)
+	out, err := MstConfigReadData(PciAddress)
+	if err != nil {
+		glog.Errorf("MellanoxBlueFieldMode(): failed to get mlx nic fw data %v", err)
+		return -1, fmt.Errorf("failed to get mlx nic fw data %v", err)
+	}
+
+	attrs := []string{internalCPUPageSupplier,
+		internalCPUEswitchManager,
+		internalCPUIbVporto,
+		internalCPUOffloadEngine,
+		internalCPUModel}
+	mstCurrentData, _ := ParseMstconfigOutput(out, attrs)
+
+	internalCPUPageSupplierstatus, exist := mstCurrentData[internalCPUPageSupplier]
+	if !exist {
+		return 0, fmt.Errorf("failed to find %s in the mstconfig output command", internalCPUPageSupplier)
+	}
+
+	internalCPUEswitchManagerStatus, exist := mstCurrentData[internalCPUEswitchManager]
+	if !exist {
+		return 0, fmt.Errorf("failed to find %s in the mstconfig output command", internalCPUEswitchManager)
+	}
+
+	internalCPUIbVportoStatus, exist := mstCurrentData[internalCPUIbVporto]
+	if !exist {
+		return 0, fmt.Errorf("failed to find %s in the mstconfig output command", internalCPUIbVporto)
+	}
+
+	internalCPUOffloadEngineStatus, exist := mstCurrentData[internalCPUOffloadEngine]
+	if !exist {
+		return 0, fmt.Errorf("failed to find %s in the mstconfig output command", internalCPUOffloadEngine)
+	}
+
+	internalCPUModelStatus, exist := mstCurrentData[internalCPUModel]
+	if !exist {
+		return 0, fmt.Errorf("failed to find %s in the mstconfig output command", internalCPUModel)
+	}
+
+	// check for DPU
+	if strings.Contains(internalCPUPageSupplierstatus, ecpf) &&
+		strings.Contains(internalCPUEswitchManagerStatus, ecpf) &&
+		strings.Contains(internalCPUIbVportoStatus, ecpf) &&
+		strings.Contains(internalCPUOffloadEngineStatus, enabled) &&
+		strings.Contains(internalCPUModelStatus, embeddedCPU) {
+		glog.V(2).Infof("MellanoxBlueFieldMode():card %s in DPU mode", PciAddress)
+		return bluefield2Dpu, nil
+	} else if strings.Contains(internalCPUPageSupplierstatus, extHostPf) &&
+		strings.Contains(internalCPUEswitchManagerStatus, extHostPf) &&
+		strings.Contains(internalCPUIbVportoStatus, extHostPf) &&
+		strings.Contains(internalCPUOffloadEngineStatus, disabled) &&
+		strings.Contains(internalCPUModelStatus, embeddedCPU) {
+		glog.V(2).Infof("MellanoxBlueFieldMode():card %s in ConnectX mode", PciAddress)
```

Comment:
I prefer to have the log happen outside this function

Reply:
I will not like to add another if outside the function. why do you prefer it outside?

Reply:
Just a preference. I expect that a function that checks the mode doesn't print to the log. The function that is asking to check for the mode could of course check. Separation of concerns imo.